WIP: split Course into CourseAnchor and CourseData

e-nastasia opened the pull request:
(this is a PR for [this](https://github.com/holochain-devcamp/learning-pathways/issues/4) issue)

The code is compiling but doesn't have any tests yet and still has a lot to refactor & change. There are also a few questions that need to be clarified in order to finish this.

e-nastasia committed:
WIP: split Course into CourseAnchor and CourseData

The code compiles, but who knows how would it work -- there are no tests
yet.
Also, it has lots of redundant parts that need to be refactored.

e-nastasia commented on dna/zomes/courses/code/src/course.rs:
```diff
@@ -93,47 +137,84 @@ pub fn course_entry_def() -> ValidatingEntryType {
-            }
-        )
+            to!( // to query this course's latest data
+                "%agent_id", // QUESTION: what should I specify here?
```
I don't understand if the `"%agend_id"` is another plain string or is this parameter specified according to some strict format. And looking at [the doc](https://docs.rs/hdk/0.0.41-alpha4/hdk/macro.to.html) doesn't help me understand how should I fill this parameter for the CourseAnchor.
@guillemcordoba @HedayatAbedijoo could you please help me with that?

guillemcordoba replied:
So the `%agent_id` refers to the entry type that you're linking to/from. In normal circumstances, you can only put entry types you have defined somewhere in your application (in our case, the entry types we have are things like "course_data" and "module" and "content"). 

But there is an exception: when you want to link directly from/to an agent public address. In that case, you put the reserved "system type" `%agent_id`. 

In our case, we need to put "course_data" there.

e-nastasia replied:
@guillemcordoba thank you for the clarification!

e-nastasia commented on dna/zomes/courses/code/src/course.rs:
```diff
@@ -259,15 +436,87 @@ pub fn add_module_to_course(
-        let mut course_entry = Course::try_from(current_course.clone())
+    let current_course_anchor = hdk::get_entry(course_address).unwrap().unwrap();
+
+    // QUESTION: why are we doing it like that?
```
and this one is more about code consistency. We're treating entry retrieval differently in `update` and `get_my_courses` methods and I just want to understand which method is more appropriate. 
@HedayatAbedijoo @guillemcordoba could you please clarify that?

guillemcordoba replied:
So here, two things: 

* I greatly discourage the use of `unwrap()`, basically anywhere. What unwrap will do is throw a panic when the result is an error or the option is none, and both cases can perfectly happen in this scenario. In the case of `Result`, always use `?` to return the error upwards and in the case of `Option` always match it or destructure it. 

* In this case, we want to fail if we don't get the course. So I think `get_as_type` it's the best option, as it's basically doing a `get_entry`, unwrapping the option and doing a `try_from` underneath, and if anything fails, returns an error.

Did we have panics in our devcamp code??? We should never write them in our applications, holochain doesn't really know what to do with those errors.

e-nastasia replied:
UPD: it has been clarified by @HedayatAbedijoo that since this is an educational app, we're just showing different ways of how something can be done.

e-nastasia replied:
UPD: it has been clarified by @HedayatAbedijoo that since this is an educational app, we're just showing different ways of how something can be done. So it's ok that it looks inconsistent.

HedayatAbedijoo replied:
I still think it is good to have both versions with comment above to show all options. but feel free to follow your taste.

e-nastasia committed:
course: refactored validations, now validating teacher in both CourseAnchor & CourseData

e-nastasia committed:
course: switched validation order to prioritize author check

If someone can't change a course, they should learn about it first, regardless
of how invalid their data is. Because this isn't something they would be
able to change and there's no reason to try and correct data only to get
the "Only teacher can ... " error after that.

guillemcordoba commented on dna/zomes/courses/code/src/course.rs:
```diff
@@ -21,119 +21,223 @@ use std::convert::TryFrom;
     pub title: String,
     pub modules: Vec<Address>,
     pub timestamp: u64,
     pub teacher_address: Address,
```
Is the teacher address correct to store with the CourseData?

e-nastasia replied:
@guillemcordoba there are 2 reasons for that:
1) teacher address improves chances of every particular CourseData entry being unique
2) since CourseData contains everything describing the course from the user perspective, we need to include teacher address as well

guillemcordoba commented on dna/zomes/courses/code/src/course.rs:
```diff
@@ -21,119 +21,223 @@ use std::convert::TryFrom;
-                    validate_course_title(&entry.title)
+                    validate_only_teacher_can_do(validation_data.sources(), &entry.teacher_address, "create")
                 },
                 EntryValidationData::Modify { new_entry, old_entry, validation_data, .. } => {
```
Can the anchor ever be modified? I don't think so...

e-nastasia replied:
@guillemcordoba you're right. That was just inertia in my thinking.

guillemcordoba commented on dna/zomes/courses/code/src/course.rs:
```diff
@@ -21,119 +21,223 @@ use std::convert::TryFrom;
+                    validate_course_title(&entry.title)
+                },
+                EntryValidationData::Modify { new_entry, old_entry, validation_data, .. } => {
+                    validate_only_teacher_can_do(validation_data.sources(), &old_entry.teacher_address, "modify")?;
```
I'd remove the teacher_address from course_data, and here in the validation I'd query the course_anchor to get the teacher_address and see if they are signing the course_data creation/modification/deletion. Here it is safe to do get_entry on the anchor since it will never change.

e-nastasia replied:
@guillemcordoba  I don't think removing the teacher_address from the course_data would bring any benefits, and querying another entry on every validation would worsen the performance (however not that much, I suppose).
Why do you think we should do that? I'm not sure I understand your reasons

guillemcordoba commented on dna/zomes/courses/code/src/course.rs:
```diff
@@ -157,82 +261,173 @@ pub fn anchor_entry_def() -> ValidatingEntryType {
-        course.timestamp,
+        course_data.teacher_address,
+        course_data.course_anchor,
+        // TODO: we shouldn't be using old timestamp here. Maybe we should receive it as a parameter for update?
```
Yes! Definetly, we should receive timestamp as a parameter to avoid update cycles.

guillemcordoba commented on dna/zomes/courses/code/src/course.rs:
```diff
@@ -157,82 +261,173 @@ pub fn anchor_entry_def() -> ValidatingEntryType {
+    // since CourseAnchor entry stays the same, we're not returning any new addresses here
+    // and since we don't have ownership of the course_address in this method, we're cloning it
+    // to comply with the return type requirements
+    Ok(course_address.clone())
```
We can also just return ()

e-nastasia replied:
@guillemcordoba the idea was to comply with the API format returning some address as result. But perhaps with this pattern we could change that (since it is very common to return never changing anchor address).

guillemcordoba commented on dna/zomes/courses/code/src/course.rs:
```diff
@@ -157,82 +261,173 @@ pub fn anchor_entry_def() -> ValidatingEntryType {
-    let course: Course = hdk::utils::get_as_type(address.clone())?;
+    // remove all CourseData for the course
+    for course_data_addr in course_data_addresses {
+        hdk::remove_entry(&course_data_addr)?;
```
Necessary? It will only add new entries to the DHT (won't actually clean up any memory, only add new things to hold and so on) and these course_data entries will be basically inaccessible once we delete the course_anchor

guillemcordoba commented on dna/zomes/courses/code/src/course.rs:
```diff
@@ -157,82 +261,173 @@ pub fn anchor_entry_def() -> ValidatingEntryType {
+    let course_anchor: CourseAnchor = hdk::utils::get_as_type(course_address.clone())?;
     for student in students {
-        hdk::remove_link(&student, &address, "student->course", "")?;
+        hdk::remove_link(&student, &course_address, "student->course", "")?;
```
Not really necessay, if we remove the anchor and the links are attached to the anchor, the links will be considered non-existent as well (need to check but I think this is automatic)

guillemcordoba commented on dna/zomes/courses/code/src/course.rs:
```diff
@@ -157,82 +261,173 @@ pub fn anchor_entry_def() -> ValidatingEntryType {
     }
-    hdk::remove_link(&course.teacher_address, &address, "teacher->courses", "")?;
+    // remove link between teacher and this course
+    hdk::remove_link(
```
This is still necessary because it does not have the removed anchor as the base.

e-nastasia replied:
@guillemcordoba I got initially confused by this comment and the comment above for `hdk::remove_link(&student, &course_address, "student->course", "")?;` which made me review links we have defined for the CourseAnchor.
And I see that both "student->courses" and "teacher->courses" don't have CourseAnchor as a base so I suppose they both should be treated the same? And therefore, should be removed explicitly.
However, we also have link "course->students" and it has CourseAnchor as base. Following what you say, this link doesn't have to be removed explicitly and that is what is already happening.
Am I understanding this correctly?

guillemcordoba commented on dna/zomes/courses/code/src/course.rs:
```diff
@@ -259,15 +454,86 @@ pub fn add_module_to_course(
-    let current_course = hdk::get_entry(course_address).unwrap().unwrap();
-    if let Entry::App(_, current_course) = current_course {
-        let mut course_entry = Course::try_from(current_course.clone())
+    let current_course_anchor = hdk::get_entry(course_address).unwrap().unwrap();
```
Much better to use option's `.ok()` and so on functions than unsafely unwrapping. This can lead to panics and we don't ever want those.

guillemcordoba reviewed (commented):
Have the tests been run?

e-nastasia commented:
Things that are left to be done here:
1. refactor `add_module_to_course` and `delete_module_from_course` moving away shared functionality related to the CourseAnchor & CourseData pattern. **UPD**: done [here](https://github.com/holochain-devcamp/learning-pathways/pull/5/commits/415bf1009a3dc2779f4db94a23c2dd12c3c152fa)
2. implement link validation for CourseAnchor: validate that we're only linking to such CourseData entries that have `course_anchor` pointing to the current CourseAnchor instance
3. implement method to convert entry address into CourseAnchor entry. Input: address of either CourseAnchor or CourseData, output: entry of CourseAnchor retrieved using input / error. Logic: if we receive CourseAnchor address, return the entry at this address. If we receive address of CourseData entry, retrieve it and follow it's `course_anchor`, return related CourseAnchor entry. If this address corresponds to neither CourseAnchor/CourseData, return an error. **UPD**: I realized that we often need both CourseAnchor & CourseData, so I wrote a method that retrieves both [here](https://github.com/holochain-devcamp/learning-pathways/pull/5/commits/f6aea018979f7c5efe2c18c1c931b75d084e5fc4)
4. implement timestamp update in `update` method. For a clear separation of concerns, it makes sense to just increment timestamp +1 in this PR and then change `update` method's interface to receive timestamp from the user in a separate PR.
5. implement backend tests (not integration ones involving UI). Some ideas on what we can test:

    5.1. `create` method results in one CourseAnchor and one CourseData that are tied by one explicit link of type `"course_anchor->course_data"` and by one implicit link (CourseData field `course_anchor` points to address of the corresponding CourseAnchor)
    5.2. `update` method results in a new CourseData entry & new link of type `"course_anchor->course_data"`
    5.3. `list` method retrieves latest CourseData corresponding for each course (this one can easily share context with 5.2. test)
6. fix the integration tests (all 9 are currently failing)

e-nastasia committed:
course: first implementation of get_anchor_and_latest_data_entries

e-nastasia committed:
refactored add_module_to_course and delete_module_from_course

Shared functionality for these methods is at:
- get_anchor_and_latest_data_entries: handles entry conversion and
retrieval;
- commit_course_data: handles saving updated entry to holochain.

e-nastasia committed:
course: refactored CRUD methods to use get_anchor_and_latest_data_entries and commit_course_data

e-nastasia commented:
> Have the tests been run?

@guillemcordoba yes, and they're failing as of now, so there will be further changes to fix that

e-nastasia committed:
course: fix for anchor-all-courses link to course_anchor

e-nastasia committed:
course_anchor: remove validation for modify

e-nastasia committed:
course: don't try to remove related CourseData entries on course deletion

e-nastasia committed:
test: more comments on test scenarios 0 and 9

e-nastasia committed:
dist: dna updates
